core/data_stream.py

```diff
@@ -0,0 +1,249 @@
+import abc
+import tempfile
+import os
+import typing
+from core.data_block import DataBlock
+
+Symbol = typing.Any
+
+
+class DataStream(abc.ABC):
+    """abstract class to represent a Data Stream
+
+    The DataStream facilitates the block interface.
+    From the interface standpoint, the two functions which are useful are:
+    - get_block(block_size) -> returns a DataBlock of the given block_size from the stream
+    - write_block(block) -> writes the block of data to the stream
+
+    The DataStream can act as a stream object for both writing and reading blocks
+    The two more useful sub-classes of the abstract class are FileDataStream and ListDataStream.
+    (see their description for more details)
+    """
+
+    @abc.abstractmethod
+    def reset(self):
+        """resets the data stream"""
+        pass
+
+    @abc.abstractmethod
+    def get_symbol(self):
+        """returns a symbol from the data stream, returns None if the stream is finished
+
+        This is an abstract method, and hence needs to be implemented by the subclasses
+        """
+        pass
+
+    def get_block(self, block_size: int) -> DataBlock:
+        """returns a block of data (of the given max size) from the stream
+
+        Args:
+            block_size (int): the size of the block of data to be returned.
+
+        Returns:
+            DataBlock:
+        """
+        # returns the next data block
+        data_list = []
+        for _ in range(block_size):
```

Comment:
I like that we have a general implementation but I would recommend doing a more efficient implementation for both lists and files (using the slicing operation and the read functions respectively). I think that's the way someone would do it naturally if not trying to do something generic.

Reply:
same for write_block

Reply:
Makes sense. As long as we don't mark this as @final there is no hurry to do this. Can just leave a TODO comment somewhere.

Reply:
done! added a NOTE and a TODO for now

Reply:
Thanks. Looks good.